client_plugin/drivers/shared/shared_driver.go

```diff
@@ -352,8 +360,78 @@ func (d *VolumeDriver) Create(r volume.Request) volume.Response {
 // Remove - removes individual volume. Docker would call it only if is not using it anymore
 func (d *VolumeDriver) Remove(r volume.Request) volume.Response {
 	log.WithFields(log.Fields{"name": r.Name}).Info("Removing volume ")
+	var msg string
+	var volRecord VolumeMetadata
+
+	// Cannot remove volumes till plugin completely initializes
+	// because we don't know if it is being used or not
+	if d.RefCounts.IsInitialized() != true {
+		msg = fmt.Sprintf(plugin_utils.PluginInitError+" Cannot remove volume %s",
+			r.Name)
+		log.Error(msg)
+		return volume.Response{Err: msg}
+	}
+
+	// Docker is supposed to block 'remove' command if the volume is used.
+	// Check on the local refcount first
+	if d.GetRefCount(r.Name) != 0 {
+		msg = fmt.Sprintf("Remove failed: Containers on this host VM are still using volume %s.",
+			r.Name)
+		log.Error(msg)
+		return volume.Response{Err: msg}
+	}
+
+	// Check status, then global refcount
+	keys := []string{
+		etcdPrefixInfo + r.Name,
+	}
+	entries, err := d.etcd.ReadVolMetadata(keys)
+	if err != nil {
+		msg = fmt.Sprintf("Remove failed: Failed to read metadata for volume %s from KV store. %v",
+			r.Name, err)
+		log.Warningf(msg)
+		return volume.Response{Err: msg}
+	}
+	// Unmarshal Info key
+	err = json.Unmarshal([]byte(entries[0].value), &volRecord)
+	if err != nil {
+		msg = fmt.Sprintf("Remove failed: Failed to unmarshal data. %v", err)
+		log.Warningf(msg)
+		return volume.Response{Err: msg}
+	}
+
+	// Test and set status to Deleting
+	if !d.etcd.CompareAndPut(etcdPrefixState+r.Name,
+		string(volStateReady),
+		string(volStateDeleting)) {
+		msg = fmt.Sprintf("Remove failed: Failed to set volume state to deleting")
+		msg += fmt.Sprintf("Containers on other host VM are still using volume %s.",
+			r.Name)
+		msg += fmt.Sprintf(" Host VMs using this volume: %s",
+			strings.Join(volRecord.ClientList, ","))
+		log.Warningf(msg)
+		return volume.Response{Err: msg}
+	}
+
+	// Delete internal volume
+	log.Infof("Attempting to delete internal volume for %s", r.Name)
+	internalVolname := internalVolumePrefix + r.Name
+	err = d.dockerd.VolumeRemove(context.Background(), internalVolname)
+	if err != nil {
+		msg = fmt.Sprintf("Failed to remove internal volume %s. Reason: %v", r.Name, err)
+		msg += fmt.Sprintf(" Check the status of the volumes belonging to driver %s", d.internalVolumeDriver)
+		log.Warningf(msg)
+		return volume.Response{Err: msg}
+	}
+
+	// Delete metadata associated with this volume
```

Comment:
Volumes are moved to deleting state prior to starting any actual deletion of metadata or internal volumes. These two deletions cannot be done simultaneously as one is an operation in etcd and the other a docker operation.

Deleting internal volume first ensures that if any failure happens, we will still have a metadata entry and hence a way to figure out that something more has to be done here.

If we delete metadata prior to deleting the volume, we will have no way of knowing that something went wrong if volume deletion failed then.

Activity number 18 in the milestones page will solve the issue by having a separate thread look for volumes that are stuck in "deleting" state and cleaning them up. Please let me know if I missed something.